src/python/pants/core/goals/deploy.py

```diff
@@ -0,0 +1,222 @@
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import logging
+from abc import ABCMeta
+from dataclasses import dataclass
+from itertools import chain
+from typing import Iterable
+
+from pants.core.goals.package import PackageFieldSet
+from pants.core.goals.publish import PublishFieldSet, PublishProcesses, PublishProcessesRequest
+from pants.engine.console import Console
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.process import InteractiveProcess, InteractiveProcessResult
+from pants.engine.rules import Effect, Get, MultiGet, collect_rules, goal_rule, rule_helper
+from pants.engine.target import (
+    FieldSet,
+    FieldSetsPerTarget,
+    FieldSetsPerTargetRequest,
+    NoApplicableTargetsBehavior,
+    Target,
+    TargetRootsToFieldSets,
+    TargetRootsToFieldSetsRequest,
+)
+from pants.engine.unions import union
+from pants.option.option_types import ArgsListOption
+
+logger = logging.getLogger(__name__)
+
+
+@union
+@dataclass(frozen=True)
+class DeployFieldSet(FieldSet, metaclass=ABCMeta):
+    """The FieldSet type for the `deploy` goal.
+
+    Union members may list any fields required to fulfill the instantiation of the `DeployProcess`
+    result of the deploy rule.
+    """
+
+
+@dataclass(frozen=True)
+class DeployProcess:
+    """A process that when executed will have the side effect of deploying a target.
+
+    To provide with the ability to deploy a given target, create a custom `DeployFieldSet` for
+    that given target and implement a rule that returns `DeployProcess` for that custom field set:
+
+    Example:
+
+        @dataclass(frozen=True)
+        class MyDeploymentFieldSet(DeployFieldSet):
+            pass
+
+        @rule
+        async def my_deployment_process(field_set: MyDeploymentFieldSet) -> DeployProcess:
+            # Create the underlying process that executes the deployment
+            process = Process(...)
+            return DeployProcess(
+                name="my_deployment",
+                process=InteractiveProcess.from_process(process)
+            )
+
+        def rules():
+            return [
+                *collect_rules(),
+                UnionRule(DeployFieldSet, MyDeploymentFieldSet)
+            ]
+
+    Use the `publish_dependencies` field to provide with a list of targets that produce packages
+    which need to be externally published before the deployment process is executed.
+    """
+
+    name: str
+    process: InteractiveProcess | None
+    publish_dependencies: tuple[Target, ...] = ()
+    description: str | None = None
+
+
+class DeploySubsystem(GoalSubsystem):
+    name = "experimental-deploy"
+    help = "Perform a deployment process."
+
+    required_union_implementation = (DeployFieldSet,)
+
+    args = ArgsListOption(
+        example="",
+        passthrough=True,
+        extra_help="Arguments to pass to the underlying tool.",
+    )
```

Comment:
I think that `args` should likely be provided by the individual backends rather than globally like this: so would remove this option for now.